src/Accounts/AuthenticationAssemblyLoadContext/AzAssemblyLoadContextBase.cs

```diff
@@ -0,0 +1,99 @@
+// ----------------------------------------------------------------------------------
+//
+// Copyright Microsoft Corporation
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+// http://www.apache.org/licenses/LICENSE-2.0
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// ----------------------------------------------------------------------------------
+
+using System;
+using System.Collections.Concurrent;
+using System.Reflection;
+using System.Runtime.Loader;
+
+namespace Microsoft.Azure.PowerShell.AuthenticationAssemblyLoadContext
+{
+    /// <summary>
+    /// Base class providing caching capability.
+    /// </summary>
+    internal abstract class AzAssemblyLoadContextBase : AssemblyLoadContext
+    {
+        private ConcurrentDictionary<string, Assembly> AssemblyCache { get; set; } =
+            new ConcurrentDictionary<string, Assembly>(StringComparer.OrdinalIgnoreCase);
+
+        protected AzAssemblyLoadContextBase(string name) : base(name) { }
+
+        protected void CacheAssembly(string name, Assembly assembly)
+        {
+            AssemblyCache.TryAdd(name, assembly);
```

Comment:
Will it be better to write a debug message is try add fails? If no, why not use `AddOrUpdate` to instead?

Reply:
@wyunchi-ms thanks for reviewing :)

>  write a debug message is try add fails

You are right, it would be better. However it would take much effort to implement because (1) we don't have a cmdlet instance so it's tricky to write debug log here and (2) we wouldn't want the assembly loading debug message to be mixed with normal debug message because there will be plenty of them and they are not useful for debugging in most cases.

> why not use AddOrUpdate

It does not make sense to cache the same assembly twice, so we don't need `AddOrUpdate` here. `TryAdd` is enough.